Add a gaussian sampler

pajotarthur opened the pull request:
With @emited, adding a C and CUDA functions for a gaussian grid sampler, similar with the one described in https://openreview.net/pdf?id=SkZXoGZC- .

ezyang commented:
@pajotarthur If you fix the merge conflicts I'll run this on CI for you.

ezyang commented:
@pytorchbot test this please

zou3519 commented:
@pytorchbot test this please

ezyang commented on .jenkins/macos-build-test.sh:
```diff
@@ -18,7 +18,7 @@ export MACOSX_DEPLOYMENT_TARGET=10.9
 export CC=clang
 # If we run too many parallel jobs, we will OOM
-export MAX_JOBS=2
+export MAX_JOBS=3
```
Haha, thanks. We might be able to bump this all the way to 4, actually, since we've got quad core machines. Or 5, to hide IO latency.

ezyang commented:
What's going on with the pybind11 submodule update?

pajotarthur commented:
I'm not really sure, I pulled from https://github.com/pybind/pybind11 manually by mistake a long time ago, I can checkout to the correct commit, but when I git submodule update, it reverts to the wrong one. Is there a workaround ? It doesn't seems to break anything though.

ezyang commented:
@pajotarthur The main reason to not update it here is to let a pybind11 submodule be its own commit, so that's easier to notice if you're reading the commit stream.

An easy way to fix the problem is to "redo" your commit:

1. `git reset HEAD~`
2. `git submodule update`
3. `git commit -C $oldcommithash`

Another way is to manually checkout the submodule to the correct commit, and then do a `git commit --amend`.

ezyang commented:
@pytorchbot test this please

ezyang commented:
The submodule still looks updated to me

ezyang commented:
BTW, you can check locally if you've done the fix properly by running `git diff origin/master` and seeing if the submodule update shows up there.

ezyang commented on test/test_nn.py:
```diff
@@ -4243,78 +4241,141 @@ def test_shape(N, C, IH, IW, H, W, padding_mode):
-                grid_cpu = Variable(torch.randn(D, N, H, W, 3).transpose(0, 1), requires_grad=True)
-                out_cpu = F.grid_sample(input_cpu, grid_cpu, padding_mode=padding_mode)
-                self.assertTrue(out_cpu.size() == torch.Size([N, C, D, H, W]))
+            def test_cpu_against_cuda(N, C, D, H, W, padding_mode):
```
The extra indent here was totally unnecessary

ezyang commented:
@pajotarthur Did @emited do a math check on the two kernel implementations?

pajotarthur commented:
Absolutely. Do you need the maths in writing ?

ezyang reviewed (approved):
Nope, that's good enough for me.

ezyang commented:
Looks like there is some sort of submodule update disaster. https://stackoverflow.com/questions/7718780/how-to-undo-git-submodule-update?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa can tell you how to fix it (you'll have to find the correct hashes)

pajotarthur commented:
I messed up my commit, I'll correct it soon.

ezyang commented:
Needs rebasing :(

goldsborough commented:
I think you did `git pull upstream master`. Please undo this and `git pull --rebase upstream master` instead

goldsborough commented:
You should probably also squash your commits into a single one

pajotarthur committed:
# This is a combination of 2 commits.
# This is the 1st commit message:

add gaussian sampler

add gaussian sampler

fix kernel_std

add init

# The commit message #2 will be skipped:

# add init

pajotarthur committed:
add gaussian sampler

pajotarthur committed:
add gaussian sampler

pajotarthur committed:
add gaussian sampler

add gaussian sampler

fix kernel_std

add init

pajotarthur committed:
correction pep8

pajotarthur committed:
fix kernel_std

pajotarthur committed:
add init

pajotarthur committed:
Merge branch 'master' of https://github.com/pajotarthur/pytorch

# Conflicts:
#	test/test_nn.py

pajotarthur committed:
Merge branch 'master' of https://github.com/pajotarthur/pytorch

ezyang commented:
@pytorchbot retest this please

ssnl committed:
Merge branch 'master' into master

weiyangfb commented:
@pajotarthur needs another rebasing now

sfotiadis commented:
Hey, fellas where is this standing right now?

ezyang reviewed (changes requested):
Unfortunately, due to our negligence, this PR has severely bitrotten. Probably the best thing to do is for someone to redo it in ATen (c.f. https://github.com/pytorch/pytorch/wiki/TH-to-ATen-porting-guide) and resubmit. Sorry :(

anderspapitto reviewed (commented):
.

xming521 commented:
Anything new?
